swarm/README.md

```diff
@@ -95,3 +98,96 @@ parallel significantly decreases this time but was not tested here.
+A full mini-broker can be deployed on any Swarm or on AWS
+using AWS's CloudFormation service.
+A template for a Docker for AWS CloudFormation deployment is included here
+in the file Docker.tmpl.
+To deploy a Docker for AWS stack, navigate to the CloudFormation service page,
+go to "Create new stack,"
+and upload the Docker template or instead navigate directly to [https://console.aws.amazon.com/cloudformation/home#/stacks/new?stackName=Docker&templateURL=https://editions-us-east-1.s3.amazonaws.com/aws/stable/Docker.tmpl](https://console.aws.amazon.com/cloudformation/home#/stacks/new?stackName=Docker&templateURL=https://editions-us-east-1.s3.amazonaws.com/aws/stable/Docker.tmpl).
+
+The following are appropriate choices for stack parameters:
+
+* Number of Swarm managers = 1
+* Number of Swarm worker nodes = 11
+* Swarm manager instance type? = m4.4xlarge
+* Agent worker instance type? = m4.4xlarge
+
+The rest can be left to the default options.
+
+Once the stack is created, navigate to "Running Instances" on the ec2 service
+page and identify the Docker-Manager machine.
+The Public DNS (IPv4) under the description of that machine gives you the
+address to which you can ssh as the user "docker" with the ssh key you
+used to generate your stack, for example:
+
+```
+$ ssh docker@ec2-99-999-9-99.us-east-2.compute.amazonaws.com
+```
+You should see a "Welcome to Docker!" message.
+Copy all scripts from this swarm directory to your manager machine:
+
+```
+$ scp * docker@ec2-99-999-9-99.us-east-2.compute.amazonaws.com
+```
+
+The deployment scripts need input files listing the nodes on which to run
+the different components of the mini-broker.
+To list the IDs and statuses of the available Docker Swarm nodes, run the
+following command:
+
+```
+$ docker node ls
+```
+
+One of these nodes will be listed as "Leader" under "Management Status."
+The deployment scripts will automatically choose this Leader node for running
+the central Kafka and Zookeeper hub.
+The other 11 nodes will be Swarm worker nodes.
+For this setup of 1+11 Swarm nodes, you can choose five nodes to each house
+the downstream Kafka+Zookeeper+MirrorMaker hubs and corresponding group of
+20 filters, five nodes to each house 20 consumers, and one node containing
+an alert producer generating 10,000 alerts per visit every 39 seconds.
+
+Copy the selected node IDs into the files `filternodes.txt`,
+`consumernodes.txt`, and `sendernodes.txt`, one ID per line.
+The deployment scripts can automatically size a mini-broker depending on the
+number of nodes listed in these files, up to the number of filter classes
+included in the alert_stream image.
+For example, to run only Filter001-Filter040, only list two nodes
+in the `filternodes.txt` and `consumernodes.txt` files.
+The Kafka and Zookeeper Docker images come from the Docker Hub
+organization 'confluentinc', and the MirrorMaker Docker image comes from the
+organization 'mtpatter'.
+The rest of the components use the alert_stream image.
```

Comment:
As discussed in the brownbag today, can you give instructions on building these appropriately from this repo?

Reply:
Using non-Docker Hub images requires setting up a private registry so that all nodes in a Swarm can access the images.  I can't seem to get this working because of restrictions I think on Docker for AWS that I do not yet understand.  NCSA has been able deploy things on Kubernetes with a private registry so someone else should be able to sort it out.  I will add a Docker file for mirrormaker so that LSST can switch to images on their Docker Hub org or use a private registry.